app/tests/ble/central/src/main.c

```diff
@@ -0,0 +1,370 @@
+    char dev[BT_ADDR_LE_STR_LEN];
+
+    bt_addr_le_to_str(addr, dev, sizeof(dev));
+    LOG_DBG("[DEVICE]: %s, AD evt type %u, AD data len %u, RSSI %i", dev, type, ad->len, rssi);
+
+    /* We're only interested in connectable events */
+    if (type == BT_GAP_ADV_TYPE_ADV_IND) {
+        bt_data_parse(ad, eir_found, (void *)addr);
+    } else if (type == BT_GAP_ADV_TYPE_ADV_DIRECT_IND) {
+        reconnect(addr);
+    }
+}
+
+static void start_scan(void) {
+    int err;
+
+    /* Use active scanning and disable duplicate filtering to handle any
+     * devices that might update their advertising data at runtime. */
+    struct bt_le_scan_param scan_param = {
+        .type = BT_LE_SCAN_TYPE_ACTIVE,
+        .options = BT_LE_SCAN_OPT_NONE,
+        .interval = BT_GAP_SCAN_FAST_INTERVAL,
+        .window = BT_GAP_SCAN_FAST_WINDOW,
+    };
+
+    err = bt_le_scan_start(&scan_param, device_found);
+    if (err) {
+        LOG_DBG("Scanning failed to start (err %d)", err);
+        return;
+    }
+
+    LOG_DBG("Scanning successfully started");
+}
+
+static void discover_conn(struct bt_conn *conn) {
+    int err;
+
+    LOG_DBG("Discovery started for conn");
+    memcpy(&uuid, BT_UUID_HIDS, sizeof(uuid));
+    discover_params.uuid = &uuid.uuid;
+    discover_params.func = discover_func;
+    discover_params.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+    discover_params.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+    discover_params.type = BT_GATT_DISCOVER_PRIMARY;
+
+    err = bt_gatt_discover(default_conn, &discover_params);
+    if (err) {
+        LOG_DBG("Discover failed(err %d)", err);
+        return;
+    }
+}
+
+static void connected(struct bt_conn *conn, uint8_t conn_err) {
+    char addr[BT_ADDR_LE_STR_LEN];
+
+    bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+    if (conn_err) {
+        LOG_DBG("Failed to connect to %s (%u)", addr, conn_err);
+
+        bt_conn_unref(default_conn);
+        default_conn = NULL;
+
+        start_scan();
+        return;
+    }
+
+    LOG_DBG("Connected: %s", addr);
+
+    if (conn == default_conn) {
+        if (bt_conn_get_security(conn) >= BT_SECURITY_L2) {
+            discover_conn(conn);
+        } else {
+            LOG_DBG("Setting the security for the connection");
+            bt_conn_set_security(conn, BT_SECURITY_L2);
+        }
+    }
+}
+
+static bool first_connect = true;
+static void pairing_complete(struct bt_conn *conn, bool bonded) { LOG_DBG("Pairing complete"); }
+
+static void do_disconnect_of_active(struct k_work *work) {
+    bt_conn_disconnect(default_conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+    if (clear_bond_on_disconnect) {
+        bt_unpair(BT_ID_DEFAULT, bt_conn_get_dst(default_conn));
+    }
+}
+
+static K_WORK_DELAYABLE_DEFINE(disconnect_work, do_disconnect_of_active);
+
+static void security_changed(struct bt_conn *conn, bt_security_t level, enum bt_security_err err) {
+    if (err > BT_SECURITY_ERR_SUCCESS) {
+        exit(1);
```

Comment:
Should we log something on security failure?

Reply:
Added